[AMDGPU] Restrict promote alloca on pointers across address spaces

sgundapa opened the pull request:
If the load/store of a pointer to stack that is not in the same address space, we restrict the promote alloca pass not to vectorize if the pointer storage sizes are different.
Example: In address space 0, pointer size is 64 bits.
               In address space 5, pointer size if 32 bits.
Casting the pointer across these address spaces is undefined behavior.
Assertion found through fuzzing.

Commit:
[AMDGPU] Restrict promote alloca on pointers across address spaces

If the load/store of a pointer to stack that is not in the same address space,
we restrict the promote alloca pass not to vectorize if the pointer storage
sizes are different.
Example: In address space 0, pointer size is 64 bits.
         In address space 5, pointer size if 32 bits.
Casting the pointer across these address spaces is undefined behavior.
Assertion found through fuzzing.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -796,6 +796,16 @@ bool AMDGPUPromoteAllocaImpl::tryPromoteAllocaToVector(AllocaInst &Alloca) {
 
+      // If the access type is a pointer, reject the address spaces with
+      // different pointer sizes.
+      // store <2 x ptr> %arg, ptr addrspace(5) %alloca - Reject.
```
I don't think this should be rejected. This is completely a valid code.

sgundapa replied:
https://llvm.org/docs/AMDGPUUsage.html#address-spaces
"Code that manipulates the stack values in other lanes of a wavefront, such as by addrspacecast-ing stack pointers to generic ones and taking offsets that reach other lanes or by explicitly constructing the scratch buffer descriptor, triggers undefined behavior when it modifies the scratch values of other lanes"

In theory, if we can define a behavior of saving a 32-bit value in to 64-bit value, this should be fine and the vice-versa will be challenging.

In the above example , if %alloca is defined by addrcast, we can impart a reasonable conversion.  
I am begining to understand the amdgpu arch, some of my assumptions could be wrong.

arsenm replied:
> %alloca is defined by addrcast, we can impart a reasonable conversion.

There are no conversions, and alloca is just the direct stack object reference.

shiltian reviewed (changes requested):
> Casting the pointer across these address spaces is undefined behavior.

Are you sure?

I don't think this PR is right. It probably just treats the symptoms instead of the root cause of the assertion.

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-loadstores.ll:
```diff
@@ -198,3 +167,39 @@ entry:
+
+; Will not vectorize because we are doing a load/store of a pointer across
+; address spaces of varying pointer sizes.
+define ptr @alloca_load_store_ptr64_full_ivec(ptr %arg) {
```
This is just a regression, there's no reason this can't be handled

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-loadstores.ll:
```diff
@@ -198,3 +167,39 @@ entry:
+  ret ptr %tmp
+}
+
+; Will not vectorize because we are doing a load/store of a pointer across
+; address spaces of varying pointer sizes.
```
The access sizes are the same. This is still valid, even if the code is doing illegal type punning. The transform should introduce casts to fix the types.

sgundapa replied:
Should the amdgpu promote alloca pass do this transformation or should this pass expect the IR to be in a state where pointer storages across address space are done after using an addrspacecast .

arsenm replied:
There's no reason to be talking about addrspacecast. This is just bytes. You do not need to know anything about how they are used or interpreted. The promotion should be done, and it's not important that these are pointer typed

arsenm replied:
Example 1 should be promoted. Example 2 could be promoted, with caveats. It's a UB out of bounds store, so it's not important to handle it optimally. The code merely needs to not crash on this case. In an ordinary optimization pipeline, it would have been deleted already. 

`%alloca = alloca i32, addrspace(5); store ptr %arg, ptr addrspace(5) %alloca` is another out of bounds store that the pass should just avoid crashing on

sgundapa replied:
Thanks Matt

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-loadstores.ll:
```diff
@@ -198,3 +167,39 @@ entry:
+  store <2 x ptr> %arg, ptr addrspace(5) %alloca, align 8
+  %tmp = load <4 x ptr addrspace(3)>, ptr addrspace(5) %alloca, align 8
+  ret <4 x ptr addrspace(3)> %tmp
+}
```
Should reproduce the same situation with one as a scalar, and some with int/fp types

sgundapa replied:
Sure

sgundapa replied:
Can you please be specific here on what you mean by scalar ?

arsenm replied:
i32 and double

sgundapa committed:
[AMDGPU] Restrict promote alloca on pointers across address spaces

If the load/store of a pointer to stack that is not in the same address space,
we restrict the promote alloca pass not to vectorize if the pointer storage
sizes are different.
Example: In address space 0, pointer size is 64 bits.
         In address space 5, pointer size if 32 bits.
Casting the pointer across these address spaces with varied pointer sizes is undefined behavior
Assertion found through fuzzing.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -674,7 +674,17 @@ static bool isSupportedAccessType(FixedVectorType *VecTy, Type *AccessTy,
+
+  // If both are pointer types, verify if they are compatible to copy across
+  // address spaces.
+  bool canCopyAcrossAddressSpaces = true;
+  if (AccessTy->isPtrOrPtrVectorTy() && VecTy->isPtrOrPtrVectorTy()) {
+    if (DL.getPointerSize(AccessTy->getPointerAddressSpace()) !=
+        DL.getPointerSize(VecTy->getPointerAddressSpace()))
+      canCopyAcrossAddressSpaces = false;
+  }
```
There is no copy across address spaces here, this check is conceptually wrong. You only need to verify the size is compatible. For the final code emission, you'll need to insert no-op casts to get the types to match

sgundapa replied:
Let me explore the no-op casts.
So <2 x ptr addrspace(5)> can be stored in to ptr addrspace (0) ?  Isn't this the undefined behavior ?

arsenm replied:
It may be illegal type punning but you don't have the context. It's just bytes

sgundapa replied:
Thanks Matt. I will try a few ways to avoid this casting assert.
Another quick question, at which point in optimization pipeline, we decide this is an illegal type punning leading to undefined behavior ?

arsenm replied:
It's not a property of the pass pipeline, it's a property of the system as a whole. The address space cast from addrspace(5) to 0 is not a no-op cast. If you reload it and use it as the wrong type, it will be an invalid pointer

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-loadstores.ll:
```diff
@@ -198,3 +198,33 @@ entry:
+define void @alloca_load_store_ptr_mixed_addrspace_ptrvec() {
+entry:
+  %A2 = alloca <2 x ptr addrspace(5)>, align 8, addrspace(5)
+  store  ptr undef, ptr addrspace(5) %A2, align 8
```
Should also test with a real value instead of an undef. A test with a constant leaf would be useful too

sgundapa committed:
Merge branch 'llvm:main' into cast_assert

sgundapa committed:
Merge branch 'llvm:main' into cast_assert

arsenm reviewed (changes requested):
Check is not correct, and test isn't using the correct alloca address space
